cpp/src/buffer/buffer.cpp

```diff
@@ -18,24 +23,37 @@ template <typename T>
 }
 }  // namespace
 
-Buffer::Buffer(std::unique_ptr<std::vector<uint8_t>> host_buffer, BufferResource* br)
+Buffer::Buffer(
+    std::unique_ptr<std::vector<uint8_t>> host_buffer,
+    BufferResource* br,
+    std::shared_ptr<Event> event
+)
     : br{br},
       size{host_buffer ? host_buffer->size() : 0},
-      storage_{std::move(host_buffer)} {
+      storage_{std::move(host_buffer)},
+      event_{size > 0 ? std::move(event) : nullptr} {
     RAPIDSMPF_EXPECTS(
         std::get<HostStorageT>(storage_) != nullptr, "the host_buffer cannot be NULL"
     );
     RAPIDSMPF_EXPECTS(br != nullptr, "the BufferResource cannot be NULL");
```

Comment:
Can't add the check for `event_` being non-null here if the `size` is non-zero because it might not be necessary.